pkg/utils/addons/addons.go

```diff
@@ -0,0 +1,166 @@
+package addons
+
+import (
+	mfc "github.com/manifestival/controller-runtime-client"
+	mf "github.com/manifestival/manifestival"
+	"github.com/tektoncd/operator/pkg/flag"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type pipelineGenerator interface {
+	generate(pipeline unstructured.Unstructured) (unstructured.Unstructured, error)
+}
+
+type dcPipeline struct {
+	environment string
+	nameSuffix  string
+}
+
+type deploymentPipeline struct {
+	environment string
+	nameSuffix  string
+}
+
+type knativePipeline struct {
+	environment string
+	nameSuffix  string
+}
```

Comment:
could they be objects of a single struct rather than separate structs.
can we just move the part which defer in the implementation into interface methods